### ISSUE TITLE
PSR-12 Support

### ISSUE THREAD

**Issue by julienfalque** [FriendsOfPHP/PHP-CS-Fixer#4502]
[PSR-12](https://www.php-fig.org/psr/psr-12/) just got approved.

I think PHP CS Fixer already supports almost all of its rules. The missing cases I could notice are the following:

---

- [5.1 if, elseif, else](https://www.php-fig.org/psr/psr-12/#51-if-elseif-else) ;
- [5.2 switch, case](https://www.php-fig.org/psr/psr-12/#52-switch-case) ;
- [5.3 while, do while](https://www.php-fig.org/psr/psr-12/#53-while-do-while) ;
- [5.4 for](https://www.php-fig.org/psr/psr-12/#54-for) ;
- [5.5 foreach](https://www.php-fig.org/psr/psr-12/#55-foreach) ;
- [5.6 try, catch, finally](https://www.php-fig.org/psr/psr-12/#56-try-catch-finally) ;

> [In control structures, ]expressions in parentheses MAY be split across multiple lines, where each subsequent line is indented at least once. When doing so, the first condition MUST be on the next line. Boolean operators between conditions MUST always be at the beginning or at the end of the line, not a mix of both.

I would create a fixer that could either:
- force boolean operators to always be at the beginning or the end of lines when a mix is found ;
- force boolean operators to always be at the beginning or the end regardless of their current position.

---

- [8. Anonymous Classes](https://www.php-fig.org/psr/psr-12/#8-anonymous-classes)

> The opening brace [of an anonymous class] MAY be on the same line as the class keyword so long as the list of implements interfaces does not wrap. If the list of interfaces wraps, the brace MUST be placed on the line immediately following the last interface.

I think this is currently covered by the rule `braces`. Not sure if we should add a new option to it or create a new fixer.

---

TODO

- [ ] support for boolean operator position in control structure expression
- [ ] support for anonymous class opening brace in case `implements` list wraps
- [ ] `@PSR12` ruleset

**Comment by kbond** [comment]
What about [this](https://www.php-fig.org/psr/psr-12/#3-declare-statements-namespace-and-import-statements) - specifically the part about spaces between use class/function/const. It this possible currently?

**Comment by dmvdbrugge** [comment]
No, the newlines between different use types is not yet possible, however there is a longstanding issue (without PR) addressing it: #3582

**Comment by dmvdbrugge** [comment]
> > When doing so, the first condition MUST be on the next line.

Is this part already covered somewhere?

As to your boolean operator fixer: I'd like the fixer to be able to do both, depending on config (always_front, always_back, when_mixed_front, when_mixed_back). It was not clear from you suggestion if you meant this, or just let it do only one of the bullet-points.

**Comment by julienfalque** [comment]
@kbond Indeed, I updated my post with PR ref given by @dmvdbrugge.

@dmvdbrugge Actually even the _"each subsequent line is indented at least once"_ isn't supported currently. I updated my post to add the rules. As they both deal with whitespace, maybe we could implement them in a single fixer?

I also updated my suggestion for the boolean operator position fixer, hope it's better now :) Thanks!

**Comment by jdufresne** [comment]
From #4568, perhaps PSR-12 should be the default configuration for v3.

**Comment by ErnestStaug** [comment]
Any update on psr-12 support?

**Comment by julienfalque** [comment]
Nope, nobody gave it a try yet.

**Comment by localheinz** [comment]
Perhaps it makes sense to create a branch that aggregates a `@PSR12` rule set, so we can see where we are and what's missing?

**Comment by julienfalque** [comment]
We can create a `@PSR12` ruleset indeed, even if it doesn't contain any specific rule *yet*. Though tracking progress regarding PSR-12 support is the exact purpose of the original post of this issue :)

**Comment by GrahamCampbell** [comment]
A provisional rule set for PSR-12 is:

```php
[
    '@PSR2' => true,
    'compact_nullable_typehint' => true,
    'declare_equal_normalize' => ['space' => 'none'],
    'function_typehint_space' => true,
    'no_empty_statement' => true,
    'no_leading_import_slash' => true,
    'no_leading_namespace_whitespace' => true,
    'no_whitespace_in_blank_line' => true,
    'return_type_declaration' => ['space_before' => 'none'],
];
```

I think there are some minor incompatibilities with PSR2, so this is not quite right to include PSR2 (I haven't looked in enough detail to list them all), so that will need to be addressed. There are also various bits and pieces from the spec not covered by the rules I listed here, most of which are listed in the OP (I say most of which because I am not certain we have them all, but I also don't know of anything missing right now).

**Comment by kubawerlos** [comment]
What about this https://github.com/KorvinSzanto/PHP-CS-Fixer/blob/feature/psr12/src/RuleSet.php#L60 ? Is this finished, abandoned or what?

**Comment by GrahamCampbell** [comment]
That other ruleset seems to do a lot of stuff that PSR-12 doesn't actually mandate. For example, the PSR-12 spec says nothing about the ordering of imports. It only mandates formatting requirements and separation of the different types. It doesn't say each type of import must be alphabetically ordered.

**Comment by GrahamCampbell** [comment]
Their braces fixer does look better set up in my example, though.

**Comment by DanieleGBX** [comment]
I'll ask Korvin next time I see him.

**Comment by pilif** [comment]
> > [anonymous classes] 

> I think this is currently covered by the rule braces. Not sure if we should add a new option to it or create a new fixer.

when comparing PHP_CodeSniffer's PSR-12 support with PHP-CS-Fixer I also noticed that PSR-12 is underspecified with regards to anonymous classes in that it doesn't specify whether parentheses should be used after the `new` keyword when there are no constructor arguments.

Their sample in the spec doesn't use any.

PHP_CodeSniffer insists on there not being any.

PHP-CS-Fixer puts them there via the `braces` rule.

```php
$instance new class {
};
```

vs.

```php
$instance new class() {
};
```

**Comment by mvorisek** [comment]
@pilif I am for the parenthesis to make it consistent with every other class instantiation.

**Comment by Jean85** [comment]
To be fair, [chapter 4]() says:

> The term “class” refers to all classes, interfaces, and traits.
> [...]
>When instantiating a new class, parentheses MUST always be present even when there are no arguments passed to the constructor.

If you read this strictly, it should apply to anonymous classes too.

**Comment by mvorisek** [comment]
> If you read this strictly, it should apply to anonymous classes too.

@pilif So please report to PHP_CodeSniffer

**Comment by SpacePossum** [comment]
the `braces` rule could be split up; taking the `must have braces` part to its own fixer; @ see https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/823

:+1: for allowing single line anonymous classes like we have `allow_single_line_closure`

**Comment by mvorisek** [comment]
@Jean85 Interesting, but if you scroll down https://www.php-fig.org/psr/psr-12/#4-classes-properties-and-methods , it seems there are not parenthesis - but there is no explanation and the example seems wrong to me, as `()` are normally needed when used some arguments needs to be passed to the constructor.

Related also with issue # 4756 where the docs shows/allows empty block (`{}`) without new line before the closing brace.

**Comment by SpacePossum** [comment]
btw., is this going from `new_with_braces` so if you don't like it you can already disable it?

**Comment by pilif** [comment]
> If you read this strictly, it should apply to anonymous classes too.

however, the sample code in https://www.php-fig.org/psr/psr-12/#8-anonymous-classes does not contain the parenteses

**Comment by Jean85** [comment]
We (as @phpfig) can release an errata to clarify better this point. We'll discuss that in our ML: https://groups.google.com/d/topic/php-fig/F0IVXx1LL_Q/discussion

**Comment by Jean85** [comment]
We're proposing a clarification with https://github.com/php-fig/fig-standards/pull/1206. The editor said that parentheses needed is the way to go.

**Comment by seyfer** [comment]
@GrahamCampbell from a proposed PSR-12 ruleset above you are missing
```
        'method_argument_space' => [
            'on_multiline' => 'ensure_fully_multiline',
        ],
```
because of
```
Argument lists MAY be split across multiple lines, where each subsequent line is indented once. When doing so, the first item in the list MUST be on the next line, and there MUST be only one argument per line.
```

**Comment by mercuriete** [comment]
can you create a ruleset called PSR12-beta or something like that in order to do testing?
I want to use it in our codebase.
thanks in advance.

**Comment by GrahamCampbell** [comment]
You can still use the ruleset in your codebase, without it being an actual ruleset. Your suggestion is a bad idea because we cannot then later delete the ruleset due to semver. We'd have to call it psr12, from the beginning, and have it somewhat close to psr-12.

**Comment by mercuriete** [comment]
ok, @GrahamCampbell 
Can you suggest how to use your provisional ruleset from commandline?
```
php-cs-fixer -fix . ruleset="???"
```

Sorry if you think that this temporary ruleset is useful because for me is not simple to use.
Do I have to patch master?

Thanks in advance.

**Comment by GrahamCampbell** [comment]
You don't need to provide a ruleset. The expected way to use php-cs-fixer is to use a config file.

**Comment by mercuriete** [comment]
@GrahamCampbell thanks I will read the docs.
😄

**Comment by GrahamCampbell** [comment]
A good example is the config file that this repo uses to fix itself (though you won't need the hacks they have at the bottom which are used to handle the case were an older version of the tool is used to execute the config file). ;)

**Comment by eclipxe13** [comment]
Looks like the rule `class_definition` is messing with PSR-12, the space after `class` on anonymous classes is removed, and there is no option to avoid this behavior.

before:
```diff
-        $instance = new class () {
+        $instance = new class() {
```

According to PSR-12:

* Anonymous Classes MUST follow the same guidelines and principles as closures...
* Closures MUST be declared with a space after the function keyword...
* When instantiating a new class, parentheses MUST always be present even when there are no arguments passed to the constructor. (previously discussed on this thread)

**Comment by julienfalque** [comment]
Thank you @eclipxe13, I updated the TODO list.

**Comment by quimcalpe** [comment]
PSR12 states regarding importing traits: 

> it MUST have a blank line after the use import statement.

```php
<?php

namespace Vendor\Package;

use Vendor\Package\FirstTrait;

class ClassName
{
    use FirstTrait;

    private $property;
}
```

I don't think that's possible right now with current rules, `class_attributes_separation` is the closest one but adds a space between every line, a space delimiting different types would be more appropriate:

```php
...
class ClassName
{
    use FirstTrait;
    use SecondTrait;

    public const C1;
    public const C2;

    private $property1;
    private $property2;
    private $property3;

    public function xxxx()
    {
        .....
    }
}
```

**Comment by GrahamCampbell** [comment]
We're missing a bullet point to make the unary spacing fixer configurable, to so PSR12 preset can use it without changing whitespace for the `!` operator.

**Comment by nickrupert7** [comment]
Hi there, outsider here wondering about the current status of this ticket? 

I see there's a PR open to add the PSR12 ruleset which was posted 2 weeks ago. Are there plans to merge it soon? I'd love to get this working with my team's projects as soon as it is

**Comment by GrahamCampbell** [comment]
I don't think it's ready yet. There are still lots of missing features from the fixers that need implementing.

**Comment by GrahamCampbell** [comment]
Many of those are listed in the OP. The repo is open to PRs if anyone wants to have a go at implementing any part of them, even if it's just writing tests for stuff.

**Comment by hamdrew** [comment]
Should the `@PSR2` ruleset be deprecated once the `@PSR12` ruleset is stable?

**Comment by Rohaq** [comment]
> 
> 
> Should the `@PSR2` ruleset be deprecated once the `@PSR12` ruleset is stable?

I don't think so - there's always going to be some poor sod out there stuck on an old codebase that might still find the ruleset useful.

**Comment by mwikala** [comment]
When is it likely that @PSR12 will become available as a rule set?

**Comment by GrahamCampbell** [comment]
Probably months. There's too much missing functionality. This repo is open to PRs. In particular, support PSR-12 style grouping of imports is a blocker for having this preset.

**Comment by caseyw** [comment]
Thanks to all who's put effort into this. Your time is appreciated.

**Comment by GrahamCampbell** [comment]
@JosephLeedy and @alexmatsak. Your downvotes show either a fundamental lack of understanding of how open source works, or that you feel other people should work for you for free. If you'd like things to move more quickly, either submit PRs implementing the missing fixers, or get a quote from someone to implement the features. Downvoting a comment which reflects reality is not helping anything, and certainly doesn't motivate people to work for free to deliver this!

**Comment by keradus** [comment]
PSR-12 ruleset merged into 2.18 and set as default for 3.0-dev
(this does not mean each PSR-12 rule is implemented - go ahead and propose the PR folks ;] )

**Comment by lsalazarm99** [comment]
I guess the way I use it in 2.18 is by adding `--rules=@PSR12` or by adding the following in a configuration file, right?

```php
...
$config->setRules(
    [
        '@PSR12' => true,
    ]
);
...
```

I'm asking because I'm doing it using the configuration file method and I still get only PSR2 rules (which are the default). For example, it is not detecting the following:

`<?php declare(strict_types=1); // There should be a blank line after "<?php"`

**Comment by BenjaminHoegh** [comment]
what's the status on this?

**Comment by GrahamCampbell** [comment]
No change.

**Comment by SpacePossum** [comment]
```
PSR12 states regarding importing traits:

    it MUST have a blank line after the use import statement.
```
this is now handled by `class_attributes_separation` with both `none` and `one` respecting different block types (not sure about `only_if_meta` someone would need to try)

**Comment by internalsystemerror** [comment]
> this is now handled by `class_attributes_separation` with both `none` and `one` respecting different block types and adding a line between each block (not sure about `only_if_meta` someone would need to try)

Using `only_if_meta` results in:
```php
    use DescribableTrait;

    protected Site $site;
    protected string $location;
    protected string $body;
    protected ?string $description = null;
```

being converted to:

```php
    use DescribableTrait;
    protected Site $site;
    protected string $location;
    protected string $body;
    protected ?string $description = null;
```

Looking for a fix to this led me here.

**Comment by SpacePossum** [comment]
could you try what happens if you use `none` or `one`? (ignoring the description but actual run, there are test for this case that show it should work)

**Comment by internalsystemerror** [comment]
> could you try what happens if you use `none` or `one`? (ignoring the description but actual run, there are test for this case that show it should work)

Using 'none', this works as expected although it adds blank lines around comments :/. I suspect the latter part may be due to some PhpDoc fixers I have included.

**Comment by movahhedi** [comment]
Rule set @PSR12[¶](https://cs.symfony.com/doc/ruleSets/PSR12.html#rule-set-psr12)
Rules that follow [PSR-12](https://www.php-fig.org/psr/psr-12/) standard.
[https://cs.symfony.com/doc/ruleSets/PSR12.html](https://cs.symfony.com/doc/ruleSets/PSR12.html)

**Comment by VincentLanglet** [comment]
For the imports section https://www.php-fig.org/psr/psr-12/#3-declare-statements-namespace-and-import-statements

> support for ordering imports by type first
> support for adding separation lines between import group

The support for ordering imports by type first is implemented https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/doc/rules/import/ordered_imports.rst#imports_order so the checkbox can be checked.

And the support for adding separation lines between import group was recently added.

But there is one missing item in this list.
The same group of import `MUST NOT contain a blank line.`

Can you add it to the list ? @julienfalque

**Comment by julienfalque** [comment]
Thanks @VincentLanglet, description updated.

**Comment by pstkweb** [comment]
I still have an issue with traits. Using `@PSR12` rule set I expects : 
```php
<?php
class MyClass
{
    use SomeTrait;
    private $em;
    // ...
}
```
To be fixed to : 
```php
<?php
class MyClass
{
    use SomeTrait;
    
    private $em;
    // ...
}
```

But it does nothing. Any clue what could I do ?

Thanks :wink:

**Comment by SpacePossum** [comment]
Hi,

Try configuring `class_attributes_separation`.
For example, with the default config:

```bash

$ php php-cs-fixer fix test.php --rules=class_attributes_separation --dry-run -vvv --diff
PHP CS Fixer 3.9.6-DEV Grand Awaiting by Fabien Potencier and Dariusz Ruminski.
PHP runtime: 7.4.24
Loaded config default from "/home/possum/work/PHP-CS-Fixer/.php-cs-fixer.dist.php".
Paths from configuration file have been overridden by paths provided as command arguments.
F                                                                                                                                                                   1 / 1 (100%)
Legend: .-no changes, F-fixed, S-skipped (cached or empty file), I-invalid file syntax (file ignored), E-error
   1) PHP-CS-Fixer/test.php (class_attributes_separation)
      ---------- begin diff ----------
```

```diff
--- /home/possum/work/PHP-CS-Fixer/test.php
+++ /home/possum/work/PHP-CS-Fixer/test.php
@@ -15,6 +15,7 @@
 class test
 {
     use SomeTrait;
+
     private $em;
     // ...
 }
```

(seems the SF set config for `class_attributes_separation` is a bit underwhelming for my liking TBH )

For more questions, please ask here:
https://gitter.im/PHP-CS-Fixer/Lobby

**Comment by a-menshchikov** [comment]
What about spaces around string binary operators (see https://www.php-fig.org/psr/psr-12/#62-binary-operators) like concatinaion?

**Comment by Wirone** [comment]
@a-menshchikov there's #6861 created recently after discussion #6849.

**Comment by ostrolucky** [comment]
PSR-PER standard has been published recently (2 weeks ago). It is an evolving standard which builds onto PSR-12.

Ref:
https://github.com/php-fig/per-coding-style/blob/master/spec.md
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/4502

**Comment by mvorisek** [comment]
@Wirone what about splitting this issue into the separate ones? I want to comment on one, but the discussion here on multitopic is quite too long to be helpful.

**Comment by Wirone** [comment]
@mvorisek I agree, that's why I unpinned this issue some time ago. First of all I believe we should focus on PER-CS now, since #6707 was merged it's now possible to work on v2. I am not sure if investing time in PSR-12 is good at this point. 

Anyway, if we want to continue working on this, we should create milestone with dedicated issues for each remaining task from this issue's checklist. @keradus what do you think about this idea? I could split this issue in the meantime.

**Comment by Jean85** [comment]
@Wirone keep in mind that PER-CS 1.0 is (intentionally) basically the same as PSR-12: https://groups.google.com/g/php-fig/c/bBh4Y-R5gxE

So you could split this up in sub-issues and keep them as valid, and just review them against the PER-CS 2.0 change list: https://github.com/php-fig/per-coding-style/releases/tag/2.0.0

**Comment by VincentLanglet** [comment]
> > [In control structures, ]expressions in parentheses MAY be split across multiple lines, where each subsequent line is indented at least once. When doing so, the first condition MUST be on the next line. Boolean operators between conditions MUST always be at the beginning or at the end of the line, not a mix of both.
> 
> We need to create a fixer that could force boolean operators to always be at the same position with two options like e.g.:
> 
> * `position` (`beginning` or `end`): choose whether the operators should be at the beginning or end of the lines;
> * `when` (`always` or `mixed`): whether the rule should always be applied or only when a mix of positions is found.
> 
> 
> TODO
> 
> * [ ]  support for boolean operator position in control structure expression

Hi @Wirone, what about adding
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.49.0/doc/rules/operator/operator_linebreak.rst
to the PSR12 ruleset with the option `only_booleans => true` ?

The rule seems pretty similar to what is asked in PSR12.

Currently it's used by Symfony Ruleset and PSR12 ruleset.

**Comment by Wirone** [comment]
@VincentLanglet yeah, it looks like `operator_linebreak` does half of the job (does not have an second mentioned config option). But for PSR-12 it should be enough, as it requires boolean operator to always be in the same position (beginning _or_ end, not a mix of both - we can enforce the default (`beginning`), and it always can be overridden when needed.

**Comment by VincentLanglet** [comment]
> yeah, it looks like operator_linebreak does half of the job (does not have an second mentioned config option)

I agree.

> But for PSR-12 it should be enough, as it requires boolean operator to always be in the same position (beginning _or_ end, not a mix of both - we can enforce the default (`beginning`), and it always can be overridden when needed.

I have made the PR https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7868#issue-2175124103 but this seemed refused at this time. I can reopen the PR but this might require some discussion with keradus.

**Comment by kenguest** [comment]
there doesn't seem to be a check re line length?

**Comment by Wirone** [comment]
> there doesn't seem to be a check re line length? ( cf [php-fig.org/per/coding-style](https://www.php-fig.org/per/coding-style/#23-lines) )

No, because:

> There MUST NOT be a hard limit on line length.

This is not something that can be automated in sensible way.

**Comment by kenguest** [comment]
Would it not then be possible to just emit a warning in that case? (Which is what PHPCodeSniffer does)

**Comment by ostrolucky** [comment]
Unlike PHPCodesniffer, php-cs-fixer doesn't have a concept of warning or error violations (non automatically fixable error), all the violations it has must be fixable. What it could do is to have some fixer which splits the code to multiple lines according some rules, like https://github.com/symplify/coding-standard?tab=readme-ov-file#linelengthfixer

**Comment by SpacePossum** [comment]
- [ ] support for removing blank lines between imports of the same group
pending in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/8171

- [ ] support for forcing blank line between import statement types (https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3582)
this has been done using the `blank_line_between_import_groups` rule

- [ ] support for anonymous class opening brace in case implements list wraps
this has been done using the `class_definition` rule
